### DISCUSSION TITLE
how to bring focused window to top ?

### DISCUSSION THREAD

**Discussion by ghost** [i3/i3#5244] [Ideas]
I'm using rofi's window switcher. Asked this question & they said it's window manager's job to sort windows, not the switcher.
Open w1, w2 & w3. All list as `w3 -> w2 -> w1`. But changing the focus between them doesn't bring any window to the top.
But strangely, moving windows to other positions, i3 starts shuffling them around. I'm not sure how i3 sorts windows, but it's too confusing. Why not bring focused window to the top & rest to the bottom, as they all should ?

**Comment by okraits** [comment]
Which layout are you using? Tiled windows should all have the same layer. Even with floating windows I can't reproduce this. Selecting a window in rofi brings it to the top.

**Reply by ghost** [reply]
Only the default tiled layout. For me it doesn't. It's very easy to reproduce this.
1. Open w1 & w2. The order in rofi, should be `w2 -> w1`.
2. Now switch from w2 to w1 in rofi. But the order doesn't change.
w1 never comes to top (atleast in my case).

**Reply by ghost** [reply]
The only way to bring w1 to top, is to swap/move the w1 position to w2 position.
[Here's my config btw](https://gitlab.com/c00per/arch/-/blob/main/.config/i3/config).

**Reply by okraits** [reply]
I'm not sure if this is an issue in i3 or if you have to change the way that rofi sorts windows.

**Reply by ghost** [reply]
I haven't configured rofi much, just gruvbox theme that came bydefault & that's it. I'm not sure why only I'm getting this issue btw.
Also... rofi devs said (in quotes), they never sort the elements/values that it parses. It's all upto the WMs.

**Reply by okraits** [reply]
Then I'm pretty sure that i3 doesn't care and thus doesn't store the order in which windows are focused. And thus it always tells rofi the order of the windows in its tree structure - which doesn't change by focusing a window.

If you care about the focus history, I suggest using an external script like https://github.com/altdesktop/i3ipc-python/blob/master/examples/i3-cycle-focus.py to change focus based on the focus history.

**Reply by Airblader** [reply]
It's possible that rofi goes by the stacking order, but that order is meaningless to i3 for tiled windows as they don't overlap.

If it isn't that, we'd have to understand how rofi lists windows to say how they're ordered.

**Reply by ghost** [reply]
Tq. Btw you said that selecting a window in rofi brings it to the top. "top" here I refer to the element in the rofi list. Not the focus.
There's no issue with focus. I was referring to the rofi window list.
@Airblader Yup. That could be possible to look into.

**Reply by okraits** [reply]
With `top` I meant the stacking layer. Because that's what is visible in i3.

**Reply by Airblader** [reply]
https://github.com/davatorium/rofi/blob/579902deff5c23cf690fc76bb9bcdca4f511b0df/source/modes/window.c#L554

rofi uses the stacking order as I suspected. But stacking order is not focus order. As far as I know rofi also doesn't claim that the windows are ordered by focus, so this is probably working as intended. You're just expecting something else.

**Reply by ghost** [reply]
> With `top` I meant the stacking layer. Because that's what is visible in i3.

Then it would be a good idea to implement focus cycling isn't it ? Would help a lot for those who open lot of containers all over the workspaces.

**Reply by okraits** [reply]
I'm pretty sure that the i3 devs will tell you to implement something like this with the IPC interface. And there are already scripts for this, look at the link I provided.

**Reply by ghost** [reply]
Okay. Will check it out. But I'm also pretty much okay with the existing defaults & I rather search for window in rofi window switcher than to rely on their positions. Just wasn't aware about IPC snippets.
